distributed/client.py

```diff
@@ -578,14 +597,16 @@ def start(self, asynchronous=None, **kwargs):
                 self._should_close_loop = True
             while not self.loop._running:
                 sleep(0.001)
+
         pc = PeriodicCallback(lambda: None, 1000, io_loop=self.loop)
         self.loop.add_callback(pc.start)
         _set_global_client(self)
-        if asynchronous:
+        self.status = 'connecting'
+
+        if self.asynchronous:
```

Comment:
The problem arises because we reuse the same client in both contexts.  Consider the following:

```python
def g():
    # I run in a worker thread
    client = get_client()
    future = client.submit(inc, 1)
    return future.result()

async def f():
    async with Client(asynchronous=True, processes=False) as c:
        await c.submit(g)
```

The client in both cases is the same.  We tell the client originally that it is operating in asynchronous mode.  However when it finds itself in the `g` task it realizes that it definitely can't be in asynchronous mode and so switches over within that context.

Ideally we would have a  function that could tell us if we "were in the event loop" but this isn't doable.  So instead we take a signal from the user (this is `asynchronous=True`) but disregard it when it's obviously false.

This approach would currently fail if we were to start a synchronous client and then try to use it asynchronously, but seems to work in all other settings.